addons/website_sale/controllers/main.py

```diff
@@ -388,6 +388,86 @@ def old_product(self, product, category='', search='', **kwargs):
         # Compatibility pre-v14
         return request.redirect(_build_url_w_params("/shop/%s" % slug(product), request.params), code=301)
 
+    def _get_product_carousel_view(self, product_product=None, product_template=None):
+        if product_product:
+            return request.env['ir.ui.view']._render_template('website_sale.shop_product_carousel', values={
+                'product': request.env['product.template'].browse(product_template),
+                'product_variant': product_product,
+            })
+        return request.env['ir.ui.view']._render_template('website_sale.shop_product_carousel', values={
+            'product': product_template,
+        })
```

Comment:
Why is one version using browse and the other one is not?
Also you can construct the values beforehand instead of having two branches calling the same function.
```suggestion
        values = {
            'product': request.env['product.template'].browse(product_template)
        }
        if product_product:
            values['product_variant'] = product_product
        return request.env['ir.ui.view']._render_template('website_sale.shop_product_carousel', values=values)
```